orchard/modules/_branch.py

```diff
@@ -6,6 +6,228 @@
+        return False
+    link_module = get_module(link_file_data, module_1['name'])
+    if not check_args(module_1, module_2, link_module):
+        return False
+    if not check_args(module_2, module_1, link_module):
+        return False
+    if not check_optionals(module_1, module_2, link_module):
+        return False
+    if not check_optionals(module_2, module_1, link_module):
+        return False
+    return True
+
+
+def insert_branch_into_path(pathstring, branchnum):
+    split_file = pathstring.split("/")
+    split_file.insert(-1, str(branchnum))
+    ret = split_file[0]
+    for step in split_file[1:]:
+        ret += '/' + step
+    return ret
+
+
+def insert_branch_into_module(module, branchnum, top_level):
+    for arg in module['arguments']:
+        if 'infile' in arg:
+            if not top_level:
+                arg['infile'] = insert_branch_into_path(arg['infile'], branchnum)
+        if 'outfile' in arg:
+            arg['outfile'] = insert_branch_into_path(arg['outfile'], branchnum)
+
+
+def insert_branch_into_config(config_file_data, branchnum, dep_struct):
+    for module in config_file_data['modules']:
+        insert_branch_into_module(module, branchnum, module_in_top_level(dep_struct, module['name']))
+
+
+def build_dep_struct(link_file_data):
+    ret = {}
+    for module in link_file_data['modules']:
+        depth = depth_level(link_file_data, module)
+        if depth not in ret:
+            ret[depth] = [module]
+        else:
+            ret[depth].append(module)
+    return ret
+
+
+def module_in_top_level(dep_struct, mod_name):
+    for module in dep_struct[1]:
+        if module['name'] == mod_name:
+            return True
+    return False
+
+
+def get_path(config_module, infile):
+    if infile:
+        return next(item for item in config_module['arguments'] if list(item.keys())[0] == 'infile')['infile']
+    else:
+        return next(item for item in config_module['arguments'] if list(item.keys())[0] == 'outfile')['outfile']
+
+
+def get_filename(pathstring):
+    split_file = pathstring.split("/")
+    return split_file[-1]
+
+
+def absolutize_paths(config_file_data):
+    for module in config_file_data['modules']:
+        pass
+        # TODO
+
+
+def branch_compare(config_file_data, branchconfig, dep_struct, link_file_data):
+    matching = []
+    for level, modules in dep_struct.items():
+        for module in modules:
+            branch_comp = get_module(branchconfig, module['name'])
+            new_comp = get_module(config_file_data, module['name'])
+            if not modules_align(new_comp, branch_comp, link_file_data):
+                return matching, False
+            matching.append(module['name'])
+    return matching, True
+
+
+def branching(config_file_data, link_file_data, workspace_path):
+    dep_struct = build_dep_struct(link_file_data)
+
+    work_path = Path(workspace_path + "/branchlog.yaml")
+    if not work_path.is_file():
+        logfile = open(work_path, 'w+')
```

Comment:
Best practice is to open a file within a context manager, so that it is auto-closed once the functionality within it is ran.

```python
with work_path.open('w+') as fh:
    ...
    yaml.safe_dump(logdata, fh, default_flow_style=False)
```

Or make sure to close the file at the end of the function.